File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> subSets = new ArrayList<>();
        List<Integer> curSet = new ArrayList<>();
        helper(0, nums, curSet, subSets);
        return subSets;    
    }

    public void helper(int i, int[] nums, List<Integer> curSet, List<List<Integer>> subSets){
        if(i>= nums.length){
           subSets.add(new ArrayList<>(curSet)); //make a copy
           return;
        }

        // decsion to include
        curSet.add(nums[i]);
        helper(i+1, nums, curSet, subSets);
        curSet.remove(curSet.size() -1);

         // decsion not to include
        helper(i+1, nums, curSet, subSets); 
    }
}

File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        subsets, curset = [], []
        self.helper(0, nums, curset, subsets)
        return subsets
    
    def helper(self, i, nums, curset, subsets):
        if i>= len(nums):
            subsets.append(curset.copy())
            return
        
        #decision to include nums[i]
        curset.append(nums[i])
        self.helper(i+1, nums, curset, subsets)
        curset.pop()

        #decision NOT to include nums[i]
        self.helper(i+1, nums, curset, subsets)